### PR TITLE
Keymaker Operation

### PR DESCRIPTION
**Pull request by myfashionhub**
Enable easily making requests to Unimatrix::Authorization/Keymaker (similar to Boxxspring::Operation)

### CODE REVIEW

File: lib/unimatrix/authorization/filters/requires_policies.rb [hunk 1]
```diff
@@ -80,7 +80,8 @@ def policies
     end
 
     def retrieve_policies( resource_name, access_token, realm )
-      if resource_name && access_token && realm
+      if resource_name && access_token
+        realm    = realm || '*'
```
Comment (jaclynj) [inline lib/unimatrix/authorization/filters/requires_policies.rb]: Why did you change this? Endpoints that use this before filter for policies should be scoped to a realm. We need to know which realm they are trying to access to see if they have permissions. A wildcard will find any permissions with that resource in any realm, which is not what we want.

**Review by jaclynj — CHANGES_REQUESTED** [review]

File: lib/unimatrix/authorization/filters/requires_policies.rb (hunk 1 above)
Comment (myfashionhub) [reply lib/unimatrix/authorization/filters/requires_policies.rb]: Because the /realms endpoint returns all policies/realms an user has access took. There is no realm uuid to check against.
This is the search query `'realm/*::regent::realms/*'`. The above is so that `requires_policies :realms` doesn't throw error.

Comment (jaclynj) [reply lib/unimatrix/authorization/filters/requires_policies.rb]: Ok I think this is fine for querying, however for creating a new realm it wouldn't be checking the right permission right? Because the way it works now, if you try to write a realm, the policies before filter will find all your permissions for every realm and if you have the write action for any of them, it will work. But if you only have access to one realm you shouldn't be able to create new ones or update other ones.

File: lib/unimatrix/authorization/railtie.rb [hunk 2]
```diff
@@ -9,12 +9,13 @@ class Railtie < Rails::Railtie
       end
     end
 
-    def retrieve_policies( resource_name, access_token, realm )
-      if resource_name && access_token && realm
-        resource  = "realm/#{ realm }::#{ ENV['APPLICATION_NAME'] }::#{ resource_name }/*"
-        params    = "resource=#{ resource }&access_token=#{ access_token }"
-        Rails.cache.fetch( Digest::SHA1.hexdigest( params ), expires_in: 1.minute ) do
```
Comment (myfashionhub) [inline lib/unimatrix/authorization/railtie.rb:16]: This didn't work because hexdigest takes string argument.

File: lib/unimatrix/authorization/filters/requires_policies.rb [hunk 3]
```diff
@@ -70,29 +69,30 @@ def self.included( controller )
...
-      response = Net::HTTP.get( uri )
-      JSON.parse( response )[ 'policies' ] rescue nil
+    def request_policies( resource_name, access_token, realm_uuid )
+      if resource_name && access_token
+        realm_uuid = realm_uuid || '*'
+
+        Unimatrix::Authorization::Operation.new( '/policies' ).where( {
+          access_token: access_token,
+          resource: "realm/#{ realm_uuid }::#{ ENV['APPLICATION_NAME'] }::#{ resource_name }/*"
+        } ).query
```
Comment (myfashionhub) [inline lib/unimatrix/authorization/filters/requires_policies.rb]: Ensure realm write policy https://github.com/bedrocketjmd/unimatrix-regent/pull/1/files

File: lib/unimatrix/authorization/filters/requires_policies.rb [hunk 4]
```diff
@@ -70,29 +69,30 @@ def self.included( controller )
       controller.extend( ClassMethods )
     end
 
-    def policies=(attributes)
+    def policies=( attributes )
       @policies = attributes
     end
 
     def policies
-      @policies ||= begin
```
Comment (jaclynj) [inline lib/unimatrix/authorization/filters/requires_policies.rb:78]: This needs to be put back to work the way it was. Removing this will break archivist policy checks in certain places, where the before action isn't called but we check policies specifically - basically anywhere that uses the "has_permission?" method.

**Review by jaclynj — CHANGES_REQUESTED** [review]

File: lib/unimatrix/authorization/filters/requires_policies.rb (hunk 4 above)
Comment (myfashionhub) [reply lib/unimatrix/authorization/filters/requires_policies.rb:78]: I think this block never worked. The controllers from Archivist that have policies have to use `requires_policies` filter. The policies are defined in the above block: 
```
controller.policies = policies
```
Both `controller_name` and `realm_uuid` are undefined here so this couldn't have worked.

Comment (jaclynj) [reply lib/unimatrix/authorization/filters/requires_policies.rb:78]: What I'm saying is that not all use the requires_policies filter. Some use the has_permission? method to check. 

For example PropertiesController#create, does not use the before filter because its a PUT to /properties with no property_id in the params. I'll admit this is kind of weird, it should be a POST to /properties/:property_id, but this will all change eventually with your regent work.

Also the ArtifactsController uses it to check if the user has create and publishing permissions - though in the artifacts controller policies would have already been set, but you may just want to check that this gem update does not break that check.

"Both controller_name and realm_uuid are undefined here so this couldn't have worked." I know this works. In archivist realm_uuid would be defined at this point and controller_name is a rails helper method. 

Perhaps this can be refactored better but without this, updating properties in portal will definitely break.

**Review by jaclynj — CHANGES_REQUESTED** [review]

**Review by jaclynj — APPROVED** [review]

File: lib/unimatrix/authorization/filters/requires_policies.rb [hunk 5]
```diff
@@ -70,29 +68,34 @@ def self.included( controller )
...
       end
     end
 
-    def request_policies( params )
-      uri = URI.parse( "#{ ENV['KEYMAKER_URL'] }/policies?#{ params }" )
-      response = Net::HTTP.get( uri )
-      JSON.parse( response )[ 'policies' ] rescue nil
+    def request_policies( resource_name, access_token, realm_uuid )
+      if resource_name && access_token
+        realm_uuid = realm_uuid || '*'
```
Comment (jaclynj) [inline lib/unimatrix/authorization/filters/requires_policies.rb:91]: The issue with writing realms I brought up here hasn't been addressed https://github.com/bedrocketjmd/unimatrix-keymaker-sdk/pull/7#discussion_r88712459

The query for ```resource: "realm/*::regent::realms/*"``` will return all policies for updating realms. So if i user has a policy to update realm 123, but not the permission to create/update all realms (*), this will still return the policy for realm 123 and they will be allowed to create a realm even though they shouldn't.

**Review by jaclynj — CHANGES_REQUESTED** [review]

File: lib/unimatrix/authorization/filters/requires_policies.rb (hunk 5 above)
Comment (myfashionhub) [reply lib/unimatrix/authorization/filters/requires_policies.rb:91]: It's checked in realms controller write action https://github.com/bedrocketjmd/unimatrix-regent/blob/master/app/controllers/realms_controller.rb (although I should add `p.realm_uuid.nil?` as well). It's kind of a special case so I didn't build it in this SDK.
